E2e cv blog post part 3 (#3484)
iterative/dvc.org: content/blog/2022-05-06-end-to-end-computer-vision-api-part-3-remote-exp-ci-cd.md

---
title:
  'End-to-End Computer Vision API, Part 3: Remote Experiments & CI/CD For
  Machine Learning'
date: 2022-05-06
description: >

>   This is the third and last part of this series of posts on how to build an

Reviewer: this description is too long for the front page

Author: Change the description

  end-to-end Computer Vision solution according to the best available MLOps
  practices. 

  In part 1, we talked about exploratory work in Jupyter Notebooks; versioning
  data in remote storage with DVC; and refactoring the code from Jupyter
  Notebooks into DVC pipeline stages.

  Part 2 talked about the process of managing experiments with DVC pipelines,
  DVCLive and Iterative Studio.

> descriptionLong: |

Reviewer: please try > instead of | in this case - its should auto-format it properly. You need | only when a list of items is needed as far as I remember.

> 
> descriptionLong: |
> 

Reviewer 2: ```suggestion
descriptionLong: |
```

>   This is the third and last part of this series of posts on how to build an end-to-end Computer Vision solution according to the best available MLOps practices. 

Reviewer: please format it (under 80 symbols) - unfortunately this is the only place if we want to have some complicated MD that we can't automate the formatting for

>   In part 1, we talked about exploratory work in Jupyter Notebooks; versioning data in remote storage with DVC; and refactoring the code from Jupyter Notebooks into DVC pipeline stages.

Reviewer: put a link? - part1

>   Part 2 talked about the process of managing experiments with DVC pipelines, DVCLive and Iterative Studio.

Reviewer: put a link part 2?

> picture: <TBD>

Reviewer 2: TODO

Author: updated with the cover image

author: alex_kim
commentsUrl: https://discuss.dvc.org/t/end-to-end-computer-vision/1178
tags:
  - Computer Vision
  - DVC
  - CML
  - Studio
  - CI/CD
  - Experiment Tracking
---

### Introduction

In this post, we’ll focus on leveraging cloud infrastructure with CML; how to

> enable automatic reporting (with metrics, plots, and other visuals) for PRs; and

Reviewer 2: what are "other visuals?" Tables? Beautifully formatted reports? Say what they are...

Author: rephrased

the eventual deployment process.

### Leveraging Cloud Resources with CI/CD and CML

> If you ever wondered if there's an easy way to quickly and easily:
> 
> 1. provision a powerful VM in the cloud (training CV models often requires
>    powerful GPUs rarely available on local machines)
> 2. submit your ML training job to it
> 3. get the results back
> 4. automatically shut down the VM without having to worry about excessive cloud
>    bills
> 
> then the answer is yes if you use the [CML library](https://cml.dev/) in

Reviewer 2: a bit hard to follow this split sentence

```suggestion
You can quickly and easily:

1. provision a powerful VM in the cloud (training CV models often requires
   powerful GPUs rarely available on local machines)
2. submit your ML training job to it
3. get the results back, and
4. automatically shut down the VM without having to pay for unused cloud resources.

All the above is possible with the [CML library](https://cml.dev/) in
```

Author: rephrased

combination with CI/CD tools like GitHub Actions or GitLab CI/CD.

> ![Continuous Integration and Deployment for Machine Learning](https://dvc.org/static/300c88b3b1b5f65753629d661cc916e5/2e49e/cicd4ml.png)

Reviewer: here and in other places - consider adding subtitles to the images

Author: done

We've configured three
[workflow files](https://github.com/iterative/magnetic-tiles-defect/tree/main/.github/workflows)
for GitHub Actions, each of which corresponds to a particular stage depending on
the project's lifecycle we are in:

1. [Workflow for experimentation and hyperparameter tuning](https://github.com/iterative/magnetic-tiles-defect/blob/main/.github/workflows/1-experiment.yaml)

   ![Workflow for experimentation and hyperparameter tuning](/uploads/images/2022-05-06/workflow_exp.png '=800')

>    In this stage, we'll be working on an experimentation git branch:

Reviewer 2: not clear. Do you mean CML will create a new branch and then open a PR? Or the user has to create a new branch? Or there's a branch already in the example repo?

Author: rephrased

   experimenting with data preprocessing, changing model architecture, tuning
   hyperparameters, etc. Once we think our experiment is ready to be run, we'll
   push our changes to a remote repository (in this case, GitHub). This push
   will trigger a CI/CD job in GitHub Actions, which in turn will:

   1. provision an EC2 virtual machine with a GPU in AWS:

      ```yaml
      ...

>       - name: deploy
>               env:
>                 REPO_TOKEN: ${{ secrets.PERSONAL_ACCESS_TOKEN }}
>               run: |
>                 cml runner \
>                     --cloud=aws \
>                     --cloud-region=us-east-1 \
>                     --cloud-type=g4dn.xlarge \
>                     --labels=cml-runner

Reviewer 2: - indentation
- reword a little to match https://cml.dev/doc/self-hosted-runners

```suggestion
      - name: Deploy runner on AWS EC2
        env:
          REPO_TOKEN: ${{ secrets.PERSONAL_ACCESS_TOKEN }}
        run: |
          cml runner \
              --cloud=aws \
              --cloud-region=us-east-1 \
              --cloud-type=g4dn.xlarge \
              --labels=cml-runner
```

      ...
      ```

   2. deploy our experiment branch to a docker container on this machine:

      ```yaml

>       ---

Reviewer 2: ```suggestion
      ...
```

Author: `restyled-io` doesn't like that https://github.com/iterative/dvc.org/pull/3515/files

      train-model:
        needs: deploy-runner
        runs-on: [self-hosted, cml-runner]
        container:
          image: iterativeai/cml:0-dvc2-base1
          options: --gpus all
        environment: cloud
        permissions:
          contents: read
          id-token: write
        steps:
          - uses: actions/checkout@v2

>           - uses: actions/setup-python@v2

Reviewer 2: I would say either:

- use `iterativeai/cml:0-dvc2-base1` and don't `setup-python`, or
- use `nvidia/cuda:11.3.1-cudnn8-runtime-ubuntu20.04` and `setup-python`, `setup-node`, `setup-cml`, and `setup-dvc`

I'd recommend the latter

Author: I opted for the former (simply removed `setup-python`) to avoid introducing many changes to the workflow file in the repo

            with:

>               python-version: '3.9'

Reviewer 2: ```suggestion
              python-version: '3.9'
      ...
```

      ```

   3. rerun the entire DVC pipeline & push metrics back to GitHub;

      ```yaml
      ...

>       - name: dvc-repro-cml
>               env:
>                 REPO_TOKEN: ${{ secrets.PERSONAL_ACCESS_TOKEN }}
>               run: |
>                 # Install dependencies
>                 pipenv install --skip-lock
>                 pipenv run dvc pull
>                 pipenv run dvc exp run
>                 pipenv run dvc push

Reviewer 2: ```suggestion
      - name: dvc-repro-cml
        env:
          REPO_TOKEN: ${{ secrets.PERSONAL_ACCESS_TOKEN }}
        run: |
          # Install dependencies
          pipenv install --skip-lock
          pipenv run dvc pull
          pipenv run dvc exp run
          pipenv run dvc push
```

>       - name: dvc-repro-cml
>               env:
>                 REPO_TOKEN: ${{ secrets.PERSONAL_ACCESS_TOKEN }}
>               run: |
>                 # Install dependencies
>                 pipenv install --skip-lock
>                 pipenv run dvc pull
>                 pipenv run dvc exp run
>                 pipenv run dvc push

Reviewer 2: ```suggestion
      - name: dvc-repro-cml
        env:
          REPO_TOKEN: ${{ secrets.PERSONAL_ACCESS_TOKEN }}
        run: |
          # Install dependencies
          pipenv install --skip-lock
          pipenv run dvc pull
          pipenv run dvc exp run
          pipenv run dvc push
```

      ...
      ```

   4. open a pull request and post a report to it that contains a table with
      metrics and model outputs on a few test images:

      ```bash
      ...
      # Open a pull request
      cml pr dvc.lock metrics.json training_metrics.json training_metrics_dvc_plots/**
      # Create CML report

>       echo "## Metrics" >> report.md
>       pipenv run dvc metrics show --md >> report.md
>       echo "## A few random test images" >> report.md
>       for file in $(ls data/test_preds/ |sort -R |tail -20); do

Reviewer 2: ```suggestion
      echo "## Metrics" > report.md
      pipenv run dvc metrics show --md >> report.md
      echo "## A few random test images" >> report.md
      for file in $(ls data/test_preds/ | sort -R | tail -20); do
```

      done
      cml send-comment --pr --update report.md
      ...
      ```

      The report structure is fully customizable. Below is an example of what
      the PR and the CML report would look like in this case. The test images
      show (from left to right) input images, ground truth masks and prediction
      masks.

      ![PR and CML report](/uploads/images/2022-05-06/pr_cml_report.png '=800')

   At this point, we can assess the results in Iterative Studio and GitHub and
   decide whether we want to accept the PR or keep experimenting.

2. [Workflow for deploying to the development environment](https://github.com/iterative/magnetic-tiles-defect/blob/main/.github/workflows/2-develop.yaml)

   ![Workflow for deploying to the development environment](/uploads/images/2022-05-06/workflow_dev.png '=800')

   Once we are happy with our model's performance on the experiment branch, we
   can merge it into the development branch. This would trigger a different
   CI/CD job that will:

   1. retrain the model if the `dev` branch contains changes not present in the
      `exp` branch. DVC will skip this stage if that's not the case. This step

>       looks almost identical to the third step in the above workflow.

Reviewer 2: link? `[third step](#some-id)`? Or better, explain:

```suggestion
      looks almost identical to step (1.3) above (rerunning the pipeline & reporting metrics on GitHub) in the above workflow.
```

   2. deploy the web REST API application (that incorporates the new model) to a
      development endpoint on Heroku:

      ```yaml

>       ---

Reviewer 2: ```suggestion
      ...
```

Author: Tried it, but `restyled-io` wants to change it back

Reviewer 2: tbh I think maybe it's better to remove all the `...` in all the snippets? Seems unneeded. Basically:

```
... # unneeded
code:
  - some
  ...  # needed, something missing in-between
  - text
... # unneeded
```

      deploy-dev-api:
        needs: train-and-push
        runs-on: ubuntu-latest
        steps:
          - uses: actions/checkout@v2
          - uses: actions/download-artifact@master
            with:
              name: model_pickle
              path: models
          - uses: akhileshns/heroku-deploy@v3.12.12
            with:
              heroku_api_key: ${{secrets.HEROKU_API_KEY}}

>               heroku_app_name: 'demo-api-mag-tiles-dev'
>               heroku_email: 'alexkim@iterative.ai'
>               team: 'iterative-sandbox'
>               usedocker: true

Reviewer 2: ```suggestion
              heroku_app_name: demo-api-mag-tiles-dev
              heroku_email: 'alexkim@iterative.ai'
              team: iterative-sandbox
              usedocker: true
      ...
```

      ```

      Now we can test our API and assess the end-to-end performance of the
      overall solution. If we pick a random test image `exp3_num_258558.jpg`,
      ![test image `exp3_num_258558.jpg`](/uploads/images/2022-05-06/exp3_num_258558.jpg '=300')

      we can send it to the endpoint using the `curl` command like this:

>       ```bash

Reviewer: bash -> dvc

Reviewer 2: ```suggestion
      ```dvc
```

      $ curl -F 'image=@data/MAGNETIC_TILE_SURFACE_DEFECTS/test_images/exp3_num_258558.jpg' \
      -v https://demo-api-mag-tiles-dev.herokuapp.com/analyze
      ```

      This will return some http-header info and the body of the response
      containing the defect segmentation mask (`0` for pixel locations without
      defects and `1` otherwise):

>       ```json

Reviewer: is it json?

Reviewer 2: probably formats better with JSON.

Author: `dvc` look ok too

   ![Workflow for deploying to the production environment](/uploads/images/2022-05-06/workflow_prod.png '=800')

   Before we merge the dev branch into the main branch, we would need to
   thoroughly test and monitor the application in the development environment. A
   good test could be duplicating real image requests to the dev endpoint (here
   it's
   [https://demo-api-mag-tiles-**dev**.herokuapp.com/analyze](https://demo-api-mag-tiles-dev.herokuapp.com/analyze))
   for some period of time and assess the quality of the returned segmentation
   masks.

   If there are no issues and we are confident in the quality of the new model,
   we can merge the development branch into the main branch of our repository.
   Again, this triggers the third CI/CD workflow that deploys the code from the
   main branch to the production API. This looks identical to the deployment
   into the development environment, except now the deployment endpoint will be
   [https://demo-api-mag-tiles-**prod**.herokuapp.com/analyze](https://demo-api-mag-tiles-prod.herokuapp.com/analyze).

## Summary

> In this three-part blog post, we described how we addressed the problem of

Reviewer 2: link to parts 1 & 2 again

building a Computer Vision Web API for defect detection. We’ve chosen this
approach because it addresses the common challenges that are shared across many
CV projects: how to version datasets that consist of a large number of small- to
medium-sized files; how to avoid triggering long-running stages of an ML
pipeline when it’s not needed for reproducibility; how to run model training
jobs on the cloud infrastructure without having to provision and manage
everything yourself; and, finally, how to track progress in key metrics when you
run many ML experiments.

We've talked about the following:

- Common difficulties when building Computer Vision Web API for defect detection
- Pros and cons of exploratory work in Jupyter Notebooks
- Versioning data in remote storage with DVC
- Moving and refactoring the code from Jupyter Notebooks into DVC pipeline
  stages
- Running experiments locally as well as remotely on cloud infrastructure with
  CML
- Tracking and visualizing experiment results in a central place with Studio

## What to Try Next

> 

Reviewer 2: ```suggestion

Missed the previous parts of this post? See [Part 1: ...](...) and [Part 2: ...](...).
Or try to:

```

Below you can see what this mask looks like.
![Output mask `exp3_num_258558_mask.png`](/uploads/images/2022-05-06/exp3_num_258558_mask.png '=300')
_Output mask `exp3_num_258558_mask.png`_

Before we merge the dev branch into the main branch, we would need to thoroughly
test and monitor the application in the development environment. A good test
could be duplicating real image requests to the dev endpoint for some period of
time and assess the quality of the returned segmentation masks.

#### 3. [Workflow for deploying to the production environment](https://github.com/iterative/magnetic-tiles-defect/blob/main/.github/workflows/3-deploy.yaml)

![Workflow for deploying to the production environment](/uploads/images/2022-05-06/workflow_prod.png '=400')
_Workflow for deploying to the production environment_

If there are no issues and we are confident in the quality of the new model, we
can merge the development branch into the main branch of our repository. Again,
this triggers the third CI/CD workflow that deploys the code from the main
branch to the production API. This looks identical to the deployment into the
development environment, except now the deployment endpoint will be

> <https://demo-api-mag-tiles-prod.herokuapp.com/analyze> (note `-prod`) .

Reviewer 2: ```suggestion
<https://demo-api-mag-tiles-prod.herokuapp.com/analyze> (note `-prod`).
```

We've talked about the following:

- Common difficulties when building Computer Vision Web API for defect detection
  ([link](https://dvc.org/blog/end-to-end-computer-vision-api-part-1-data-versioning-and-ml-pipelines#introduction))
- Pros and cons of exploratory work in Jupyter Notebooks
  ([link](https://dvc.org/blog/end-to-end-computer-vision-api-part-1-data-versioning-and-ml-pipelines#proof-of-concept-in-jupyter-notebooks))
- Versioning data in remote storage with DVC
  ([link](https://dvc.org/blog/end-to-end-computer-vision-api-part-1-data-versioning-and-ml-pipelines#data-versioning))
- Moving and refactoring the code from Jupyter Notebooks into DVC pipeline
  stages
  ([link](https://dvc.org/blog/end-to-end-computer-vision-api-part-1-data-versioning-and-ml-pipelines#refactoring-jupyter-code-into-an-ml-pipeline))
- Experiment management and versioning
  ([link](https://dvc.org/blog/end-to-end-computer-vision-api-part-2-local-experiments#experiment-management))
- Visualization of experiments and collaboration in Iterative Studio
  ([link](https://dvc.org/blog/end-to-end-computer-vision-api-part-2-local-experiments#collaboration-and-reporting-with-iterative-studio))

> - Remote experiments, CI/CD, and production deployment (this post)

Reviewer 3: Missing link

Author: Oh, that’s on purpose. I didn’t want to link from the post to itself.

Reviewer 3: Oh, gotcha.

_Workflow for deploying to the development environment_ Once we are happy with
our model's performance on the experiment branch, we can merge it into the
development branch. This would trigger a different CI/CD job that will:

a) retrain the model if the `dev` branch contains changes not present in the
`exp` branch. DVC will skip this stage if that's not the case. This step looks
almost identical to step (1.c) above (rerunning the pipeline & reporting metrics
on GitHub) in the above workflow.

b) deploy the web REST API application (that incorporates the new model) to a
development endpoint on Heroku:

```yaml
deploy-dev-api:
  needs: train-and-push
  runs-on: ubuntu-latest
  steps:
    - uses: actions/checkout@v2
    - uses: actions/download-artifact@master
      with:
        name: model_pickle
        path: models
    - uses: akhileshns/heroku-deploy@v3.12.12
      with:
        heroku_api_key: ${{secrets.HEROKU_API_KEY}}
        heroku_app_name: demo-api-mag-tiles-dev
        heroku_email: 'alexkim@iterative.ai'
        team: iterative-sandbox
        usedocker: true
```

> The development endpoint is now accessible at:
> 
> [https://demo-api-mag-tiles-dev.herokuapp.com/analyze](https://demo-api-mag-tiles-dev.herokuapp.com/analyze)

Reviewer 2: ```suggestion
The development endpoint is now accessible at
<https://demo-api-mag-tiles-dev.herokuapp.com/analyze>
```

plt.imsave(f'{file_path.stem}_mask.png', pred, cmap=cm.gray)
```

Below you can see what this mask looks like.
![Output mask `exp3_num_258558_mask.png`](/uploads/images/2022-05-06/exp3_num_258558_mask.png '=300')
_Output mask `exp3_num_258558_mask.png`_

Before we merge the dev branch into the main branch, we would need to thoroughly
test and monitor the application in the development environment. A good test
could be duplicating real image requests to the dev endpoint for some period of
time and assess the quality of the returned segmentation masks.

#### 3. [Workflow for deploying to the production environment](https://github.com/iterative/magnetic-tiles-defect/blob/main/.github/workflows/3-deploy.yaml)

![Workflow for deploying to the production environment](/uploads/images/2022-05-06/workflow_prod.png '=400')
_Workflow for deploying to the production environment_

If there are no issues and we are confident in the quality of the new model, we
can merge the development branch into the main branch of our repository. Again,
this triggers the third CI/CD workflow that deploys the code from the main
branch to the production API. This looks identical to the deployment into the

> development environment, except now the deployment endpoint will be:
> 
> [https://demo-api-mag-tiles-prod.herokuapp.com/analyze](https://demo-api-mag-tiles-prod.herokuapp.com/analyze)

Reviewer 2: ```suggestion
development environment, except now the deployment endpoint will be
<https://demo-api-mag-tiles-prod.herokuapp.com/analyze>
```
